### PR TITLE
Replace monkey patching with dynamic class generation

### PR DESCRIPTION
**Pull request by nanthony21**
The current behavior of pycromanager is that it represents java objects with the `JavaObjectShadow` class and then monkey patches each instance of `JavaObjectShadow` to match the signature of the java class.

The PR makes it so that rather than monkey patching, we generate a subclass of `JavaObjectShadow` which is named based on the original java class name, e.g. 
```
>>>type(bridge.get_studio())
        pycromanager.core.org_micromanager_internal_MMStudio.
```

This will fix some confusing behavior, for example right now this statement: `type(bridge.get_studio()) is type(bridge.get_core)` will evaluate as `True` since they are both just instances of `JavaObjectShadow`. With this PR python will be able to recognize that CMMCore and MMStudio are distinct types.

A further benefit is that rather than parsing the java API of an class each time a new `JavaShadowObject` is created in Python, we only need to generate each shadow class once. This is handled by a new `JavaObjectFactory` class. This opens the door for an enhancement I propose where java API is only transmitted once for each class as part of a separate message than the `serialized_object` messages, in some cases this could greatly reduce socket traffic.

### CODE REVIEW

**Comment by henrypinkard** [comment]
Great idea! Should `JavaObjectFactory` be `_JavaObjectFactory` since it is internal?

**Comment by nanthony21** [comment]
Good point, it's been renamed

File: pycromanager/core.py [hunk 1]
```diff
@@ -132,6 +133,9 @@ def __init__(self, port=_DEFAULT_PORT, convert_camel_case=True, debug=False):
                             '\nJava ZMQ server version: {}\nPython client expected version: {}'.format(reply_json['version'],
                                                                                            self._EXPECTED_ZMQ_SERVER_VERSION))
 
+    def get_class(self, serialized_object) -> typing.Type[JavaObjectShadow]:
```
Comment (henrypinkard) [inline pycromanager/core.py:136]: `typing.Type[JavaObjectShadow]` causes error in python3.6

File: pycromanager/core.py [hunk 2]
```diff
@@ -215,36 +218,68 @@ def get_studio(self):
...
 
+class _JavaClassFactory:
+    """
+    This class is responsible for generating subclasses of JavaObjectShadow. Each generated class is kept in a `dict`.
+    If a given class has already been generate once it will be returns from the cache rather than re-generating it.
+    """
+    def __init__(self):
+        self.classes = {}
+
+    def create(self, serialized_obj: dict, convert_camel_case: bool = True) -> typing.Type[JavaObjectShadow]:
```
Comment (henrypinkard) [inline pycromanager/core.py:229]: `typing.Type[JavaObjectShadow]` causes error in python3.6

**Comment by henrypinkard** [comment]
The two lines above are marked as errors in in PyCharm when using python 3.6. However, after downloading 3.7 (but still using 3.6) the errors go away. It appears that `__annotations__` was added to `future` in 3.7, which might explain this. Maybe it would be good to remove the annotations there since they aren't essential (I think)?

**Comment by nanthony21** [comment]
Should be OK for python 3.6 now. Do you have an idea of which versions you want to make sure this is compatible with?

**Comment by henrypinkard** [comment]
Maybe >= 3.6? What do you think?

**Comment by nanthony21** [comment]
That seems reasonable.

I don't think that the regression to using `exec` is going to play nicely with this PR. I'll switch to 3.6 and see if I can solve the `TaggedImage.pix` issue.

**Comment by nanthony21** [comment]
Good to go

**Pull request merged**